README.md
# (MODX)EvolutionCMS.snippets.ddMenuBuilder

Simple and flexible template-driven menu builder.
Initially inspired by combination of the Wayfinder and Ditto advantages with significant code simplification.


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.ru/modx/ddtools) >= 0.59


## Installation


### Using [(MODX)EvolutionCMS.libraries.ddInstaller](https://github.com/DivanDesign/EvolutionCMS.libraries.ddInstaller)

Just run the following PHP code in your sources or [Console](https://github.com/vanchelo/MODX-Evolution-Ajax-Console):

```php
//Include (MODX)EvolutionCMS.libraries.ddInstaller
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddInstaller/require.php'
);

//Install (MODX)EvolutionCMS.snippets.ddMenuBuilder
\DDInstaller::install([
	'url' => 'https://github.com/DivanDesign/EvolutionCMS.snippets.ddMenuBuilder',
	'type' => 'snippet'
]);
```

* If `ddMenuBuilder` is not exist on your site, `ddInstaller` will just install it.
* If `ddMenuBuilder` is already exist on your site, `ddInstaller` will check it version and update it if needed.


### Manually


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddMenuBuilder`.
2. Description: `<b>2.2</b> Simple and flexible template-driven menu builder.`.
3. Category: `Core → Navigation`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddMenuBuilder_snippet.php` file from the archive.


#### 2. Elements → Manage Files:

1. Create a new folder `assets/snippets/ddMenuBuilder/`.
2. Extract the archive to the folder (except `ddMenuBuilder_snippet.php`).


## Parameters description


### Data provider parameters

Providers get documents data to output.

* `provider`
	* Desctription: Name of the provider that will be used to fetch documents.
	* Valid values:
		* `'parent'`
		* `'select'`
	* Default value: `'parent'`
	
* `providerParams`
	* Desctription: Parameters to be passed to the provider.
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormatted` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as native PHP object or array (e. g. for calls through `\DDTools\Snippet::runSnippet` or `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —


#### Providers → Parent (``&provider=`parent` ``)

Select children documents from required parent(s).

* `providerParams->parentIds`
	* Desctription: Parent IDs — the starting points for the menu. Specify '0' to start from the site root.
	* Valid values:
		* `array`
		* `stringCommaSepareted`
	* Default value: `'0'`
	
* `providerParams->parentIds[i]`
	* Desctription: Parent ID.
	* Valid values: `integerDocumentID`
	* **Required**
	
* `providerParams->depth`
	* Desctription: The depth of documents to build the menu.
	* Valid values: `integer`
	* Default value: `1`


#### Providers → Select (``&provider=`select` ``)

Just output selected documents.

* `providerParams->ids`
	* Desctription: Document IDs.
	* Valid values:
		* `array`
		* `stringCommaSepareted`
	* **Required**
	
* `providerParams->ids[i]`
	* Desctription: Document IDs.
	* Valid values: `integerDocumentID`
	* **Required**


### General parameters

* `sortDir`
	* Desctription: The sorting direction (by `menuindex` field).
	* Valid values:
		* `'ASC'`
		* `'DESC'`
	* Default value: `'ASC'`
	
* `showPublishedOnly`
	* Desctription: Show only published documents.
	* Valid values:
		* `0`
		* `1`
	* Default value: `1`
	
* `showInMenuOnly`
	* Desctription: Show only documents visible in the menu.
	* Valid values:
		* `0`
		* `1`
	* Default value: `1`


### Template parameters

* `templates`
	* Desctription: Templates.  
		Placeholders available in all templates:
		* `[+id+]`
		* `[+menutitle+]` — will be equal to `[+pagetitle+]` if empty.
		* `[+pagetitle+]`
		* `[+published+]`
		* `[+isfolder+]`
		* `[+totalAllChildren+]`
		* `[+totalThisLevelChildren+]`
		* `[+level+]`
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormatted` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as native PHP object or array (e. g. for calls through `\DDTools\Snippet::runSnippet` or `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —


#### Item templates

* `templates->item`
	* Desctription: The menu item template.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:<li><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a></li>'`
	
* `templates->itemHere`
	* Desctription: The menu item template for the current document.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:<li class="active"><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a></li>'`
	
* `templates->itemActive`
	* Desctription: The menu item template for a document which is one of the parents to the current document when the current document doesn't displayed in the menu (e. g. excluded by the `depth` parameter).
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->itemHere`
	
* `templates->itemUnpub`
	* Desctription: The menu item template for unpublished document (when `showPublishedOnly` == `0`).
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->item`
	
* `templates->itemUnpubActive`
	* Desctription: The menu item template for unpublished document which is one of the parents to the current document when the current document doesn't displayed in the menu (e. g. excluded by the `depth` parameter).
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->itemActive`


#### Parent item templates

* `templates->itemParent`
	* Desctription: The menu item template for documents which has a children displayed in menu.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:<li><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a><ul>[+children+]</ul></li>'`
	
* `templates->itemParentHere`
	* Desctription: The menu item template for the current document when it has children displayed in menu.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:<li class="active"><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a><ul>[+children+]</ul></li>'`
	
* `templates->itemParentActive`
	* Desctription: The menu item template for a document which has the current document as one of the children.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->itemParentHere`
	
* `templates->itemParentUnpub`
	* Desctription: The menu item template for unpublished documents which has a children displayed in menu.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->itemParent`
	
* `templates->itemParentUnpubActive`
	* Desctription: The menu item template for an unpublished document which has the current document as one of the children.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: = `templates->itemParentActive`


#### Outer template

* `templates->outer`
	* Desctription: Wrapper template.  
		Available placeholders:
		* `[+children+]` — Generated HTML with all items.
	* Valid values:
		* `stringChunkName`
		* `string` — use inline templates starting with `@CODE:`
	* Default value: `'@CODE:<ul>[+children+]</ul>'`
	
* `placeholders`
	* Desctription:
		Additional data has to be passed into the `templates->outer`.  
		Arrays are supported too: `some[a]=one&some[b]=two` => `[+some.a+]`, `[+some.b+]`; `some[]=one&some[]=two` => `[+some.0+]`, `[some.1]`.
	* Valid values:
		* `stringJsonObject` — as [JSON](https://en.wikipedia.org/wiki/JSON)
		* `stringHjsonObject` — as [HJSON](https://hjson.github.io/)
		* `stringQueryFormatted` — as [Query string](https://en.wikipedia.org/wiki/Query_string)
		* It can also be set as native PHP object or array (e. g. for calls through `\DDTools\Snippet::runSnippet` or `$modx->runSnippet`):
			* `arrayAssociative`
			* `object`
	* Default value: —


## Examples

All examples are written using [HJSON](https://hjson.github.io/), but if you want you can use vanilla JSON instead.


### Providers → Parent

```html
[[ddMenuBuilder?
	&provider=`parent`
	&providerParams=`{
		parentId: 1
		depth: 2
	}`
]]
```


### Providers → Select

```html
[[ddMenuBuilder?
	&provider=`select`
	&providerParams=`{
		ids: [
			1
			2
			3
		]
	}`
]]
```


### Pass additional data into outer chunk (the `placeholders` parameter)

```html
[[ddMenuBuilder?
	&templates=`{
		outer:
			'''
			@CODE:<ul class="[+class+]">[+children+]</ul>[+somePlaceholder2+]
			'''
	}`
	&placeholders=`{
		class: someClass
		somePlaceholder2: <p>Some value for placeholder.</p>
	}`
]]
```


### Using Query string instead of JSON

JSON syntax is more clear than Query string, but sometimes it's not convenient. For example, if you want to pass JSON string as string.

```html
[[ddMenuBuilder?
	&provider=`parent`
	&providerParams=`parentId=1&depth=2`
	&templates=`outer=general_nav`
	&placeholders=`pladeholder1={"someName": "someValue"}&pladeholder2={"name": "John"}`
]]
```


### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
//Include (MODX)EvolutionCMS.libraries.ddTools
require_once(
	$modx->getConfig('base_path') .
	'assets/libs/ddTools/modx.ddtools.class.php'
);

//Run (MODX)EvolutionCMS.snippets.ddMenuBuilder
\DDTools\Snippet::runSnippet([
	'name' => 'ddMenuBuilder',
	'params' => [
		'providerParams' => [
			'parentId' => 1
		]
	]
]);
```


## Links

* [Home page](http://code.divandesign.ru/modx/ddmenubuilder)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddmenubuilder)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.snippets.ddMenuBuilder)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddMenuBuilder changelog


## Version 2.2 (2023-05-05)
* \+ Parameters → `providerParams`, `templates`, `placeholders`: Can also be set as [HJSON](https://hjson.github.io/) or as a native PHP object/array.
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ README:
	* \+ Installation → Using (MODX)EvolutionCMS.libraries.ddInstaller.
	* \+ Links.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.59 is required.


## Version 2.1.1 (2020-04-12)
* \+ README: Improvements.
* \+ README_ru.


## Version 2.1 (2020-03-07)
* \+ Snippet: All templates has the following placeholders:
	* \+ `[+totalAllChildren+]` — total number of displayed children at all levels.
	* \+ `[+totalThisLevelChildren+]` — total number of displayed immediate children.
	* \+ `[+level+]` — item level in menu.
* \* `\ddMenuBuilder`
	* \+ `\ddMenuBuilder::generate`:
		* \+ Return the following 2 counters:
			* \+ `$result['totalAll']` — total number of items displayed at all levels.
			* \+ `$result['totalThisLevel']` — total number of items displayed at this level.
		* \+ Added the parameter `$params->level` for internal use only.
		* \* Returns `stdClass` instead of `arrayAssociative`.
	* \* `\ddMenuBuilder::prepareProviderParams`:
		* \+ The `$params->providerParams` parameter can be set as `stdClass` too.
		* \* Returns `stdClass` instead of `arrayAssociative`.
	* \* `\ddMenuBuilder::$templates`: Now it's `stdClass`.
* \+ Composer.json.
* \+ CHANGELOG: Small improvements.
* \+ README:
	* \+ Requires.
	* \+ Documentation → Installation.


## Version 2.0 (2019-06-13)
* \* Attention! Backward compatibility is broken!
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.24.1 is required.
* \* Template parameters refactoring.
* \+ Added templates for unpublished items:
	* \+ `$templates['itemUnpub']` — The menu item template for unpublished document. Default: `$templates['item']`.
	* \+ `$templates['itemUnpubActive']` — The menu item template for unpublished document which is one of the parents to the current document when the current document doesn't displayed in the menu (e. g. excluded by the `depth` parameter). Default: `$templates['itemActive']`.
* \* `\ddTools::$modx` is used instead of global `$modx`.
* \* Fixed an error when docs that must be hidden will be showed.
* \* Refactoring, other small changes.


## Version 1.13b (2018-10-17)
* \* Attention! PHP >= 5.6 is required.
* \* Snippet:
	* \* Wrong type of `providerParams` was fixed.
* \* `\ddMenuBuilder`:
	* \* Small refactoring.
	* \* Optimization:
		* \- `\ddMenuBuilder::generate`: Redudnand `array_merge` removed,
		* \- `\ddMenuBuilder::generate`: Убран проход по всем документам в дереве который определял где находится активный документ.


## Version 1.12 (2017-08-30)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.20 is required.
* \* Menu item active status is no logner depends on the `show_in_menu` children flag.
* \+ Added JSON format support for the `providerParams` and `placeholders` parameters.


## Version 1.11 (2016-11-25)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.16.1 is required.
* \+ Added an ability to pass ids of the selected documents to output.
* \* Short array syntax is used because it's more convenient.
* \* `\ddMenuBuilder`:
	* \* Unpublished docs will be used if needed.
	* \* `\ddMenuBuilder::generate`:
		* \* Now takes custom `where` clauses instead of parent id.
		* \* Refactoring parameters style.
* \* Other minor changes.


## Version 1.10 (2016-09-12)
* \* Attention! MODXEvo >= 1.1 is required.
* \+ Added an ability to pass additional data into a `tpls_outer` template (see the `placeholders` parameter).
* \+ Added support of `@CODE:` keyword prefix in the snippet templates.


## Version 1.9 (2015-12-28)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.15 is required.
* \* Snippet:
	* \* Вместо прямого обращения к полю `$modx->config` используется метод `$modx->getConfig`.
	* \* Следующие параметры были переименованы (старые имена поддерживаются, но не рекомендуются к использованию):
		* \* `tplRow` → `tpls_item`.
		* \* `tplHere` → `tpls_itemHere`.
		* \* `tplActive` → `tpls_itemActive`.
		* \* `tplParentRow` → `tpls_itemParent`.
		* \* `tplParentHere` → `tpls_itemParentHere`.
		* \* `tplParentActive` → `tpls_itemParentActive`.
		* \* `tplUnpubParentRow` → `tpls_itemParentUnpub`.
		* \* `tplUnpubParentActive` → `tpls_itemParentUnpubActive`.
		* \* `tplWrap` → `tpls_outer`.
	* \* Параметр `tpls_itemParentHere` по умолчанию равен `<li class="active"><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a><ul>[+children+]</ul></li>` (значение по умолчанию больше не зависит от параметра `tpls_itemParent`). Решение неоднозначное, подумать.
* \* `\ddMenuBuilder` обновлён до 2.0:
	* \* Теперь это обычный объект, поля и методы не статические.
	* \* Публичный только метод `generate`, остальные поля и методы приватные.
	* \- Удалено поле `$table`, вместо него используется `\ddTools::$tables['site_content']`.
	* \* Поле `\ddMenuBuilder::$id` переименовано в `\ddMenuBuilder::$hereDocId`.
	* \+ Добавлены значения по умолчанию для полей `sortDir` и `where`.
	* \+ Значения шаблонов по умолчанию хранятся в поле `\ddMenuBuilder::$templates`.
	* \* Переименованы шаблоны.
	* \+ Добавлен конструктор.
	* \* Обработка параметров `showPublishedOnly`, `showInMenuOnly` и формирование SQL-условия вынесены из сниппета в конструктор класса `\ddMenuBuilder`.
	* \* Обработка значений шаблонов по умолчанию вынесена из сниппета в конструктор класса `\ddMenuBuilder`.
	* \* Подключение библиотеки `modx.ddTools` вынесено в конструктор.
	* \* Вместо прямого обращения к полю `$modx->config` используется метод `$modx->getConfig`.
	* \* Файл `assets/snippets/ddMenuBuilder/ddmenubuilder.class.php` переименован в `assets/snippets/ddMenuBuilder/ddMenuBuilder.class.php`.



src/Snippet.php
<?php
namespace ddMenuBuilder;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '2.2.0',
		
		$params = [
			//Defaults
			'provider' => 'parent',
			'providerParams' => [],
			
			'sortDir' => null,
			'showPublishedOnly' => null,
			'showInMenuOnly' => null,
			
			'templates' => [
				'outer' => '@CODE:<ul>[+children+]</ul>'
			],
			'placeholders' => [],
		],
		
		$paramsTypes = [
			'providerParams' => 'objectStdClass',
			'templates' => 'objectStdClass',
			'placeholders' => 'objectStdClass',
		]
	;
	
	/**
	 * prepareParams
	 * @version 1.0 (2023-03-10)
	 *
	 * @param $this->params {stdClass|arrayAssociative|stringJsonObject|stringHjsonObject|stringQueryFormatted}
	 *
	 * @return {void}
	 */
	protected function prepareParams($params = []){
		//Call base method
		parent::prepareParams($params);
		
		$this->params->templates->outer = \ddTools::$modx->getTpl($this->params->templates->outer);
	}
	
	/**
	 * run
	 * @version 1.0 (2023-05-05)
	 * 
	 * @return {string}
	 */
	public function run(){
		$ddMenuBuilder_params = (object) [
			'templates' => $this->params->templates
		];
		
		//Направление сортировки
		if (!empty($this->params->sortDir)){
			$ddMenuBuilder_params->sortDir = $this->params->sortDir;
		}
		//По умолчанию будут только опубликованные документы
		if (!empty($this->params->showPublishedOnly)){
			$ddMenuBuilder_params->showPublishedOnly = $this->params->showPublishedOnly;
		}
		//По умолчанию будут только документы, у которых стоит галочка «показывать в меню»
		if (!empty($this->params->showInMenuOnly)){
			$ddMenuBuilder_params->showInMenuOnly = $this->params->showInMenuOnly;
		}
		
		$ddMenuBuilder = new \ddMenuBuilder\Main($ddMenuBuilder_params);
		
		//Генерируем меню
		$resultObject = $ddMenuBuilder->generate(
			$ddMenuBuilder->prepareProviderParams([
				//Parent by default
				'provider' => $this->params->provider,
				'providerParams' => $this->params->providerParams
			])
		);
		
		return \ddTools::parseText([
			'text' => $this->params->templates->outer,
			'data' => \DDTools\ObjectTools::extend([
				'objects' => [
					$this->params->placeholders,
					[
						'children' => $resultObject->outputString,
						'totalAllChildren' => $resultObject->totalAll,
						'totalThisLevelChildren' => $resultObject->totalThisLevel,
					]
				]
			])
		]);
	}
}

ddMenuBuilder.php
<?php
/**
 * ddMenuBuilder
 * @version 1.13b (2018-10-17)
 * 
 * @desc Fresh, simple and flexible template-driven menu builder. Initially inspired by combination of the Wayfinder and Ditto advantages with significant code simplification.
 * 
 * @uses PHP >= 5.6.
 * @uses MODXEvo >= 1.1.
 * @uses MODXEvo.library.ddTools >= 0.20.
 * 
 * Data provider parameters:
 * @param $provider {'parent'|'select'} — Name of the provider that will be used to fetch documents. Default: 'parent'.
 * @param $providerParams {stirng_json|string_queryFormated} — Parameters to be passed to the provider. The parameter must be set as JSON (https://en.wikipedia.org/wiki/JSON) or Query string (https://en.wikipedia.org/wiki/Query_string).
 * When $provider == 'parent' =>
 * @param $providerParams['parentIds'] {array|string_commaSepareted} — Parent IDs — the starting points for the menu. Specify '0' to start from the site root. Default: '0'.
 * @param $providerParams['parentIds'][i] {integer_documentID} — Parent ID. @required
 * @param $providerParams['depth'] {integer} —  The depth of documents to build the menu. Default: 1.
 * @example &providerParams=`{"parentId": 1, "depth": 2}`.
 * @example &providerParams=`parentId=1&depth=2`.
 * When $provider == 'select' =>
 * @param $providerParams['ids'] {array|string_commaSepareted} — Document IDs. @required
 * @param $providerParams['ids'][i] {integer_documentID} — Document ID. @required
 * @example &providerParams=`{"ids": [1, 2, 3]}`.
 * @example &providerParams=`ids=1,2,3`.
 * 
 * General parameters:
 * @param $sortDir {'ASC'|'DESC'} — The sorting direction (by “menuindex” field). Default: 'ASC'.
 * @param $showPublishedOnly {0|1} — Show only published documents. Default: 1.
 * @param $showInMenuOnly {0|1} — Show only documents visible in the menu. Default: 1.
 * 
 * Template parameters:
 * All templates can be set as chunk name or code via “@CODE:” prefix.
 * Placeholders available in all templates: [+id+], [+menutitle+] (will be equal to [+pagetitle+] if empty), [+pagetitle+], [+published+], [+isfolder+].
 * @param $tpls_item {string_chunkName|string} — The menu item template. Default: '<li><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a></li>'.
 * @param $tpls_itemHere {string_chunkName|string} — The menu item template for the current document. Default: '<li class="active"><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a></li>'.
 * @param $tpls_itemActive {string_chunkName|string} — The menu item template for a document which is one of the parents to the current document when the current document doesn't displayed in the menu (e. g. excluded by the “depth” parameter). Default: $tpls_itemHere.
 * 
 * @param $tpls_itemParent {string_chunkName|string} — The menu item template for documents which has a children displayed in menu. Default: '<li><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a><ul>[+children+]</ul></li>';.
 * @param $tpls_itemParentHere {string_chunkName|string} — The menu item template for the current document when it has children displayed in menu. Default: '<li class="active"><a href="[~[+id+]~]" title="[+pagetitle+]">[+menutitle+]</a><ul>[+children+]</ul></li>'.
 * @param $tpls_itemParentActive {string_chunkName|string} — The menu item template for a document which has the current document as one of the children. Default: $tpls_itemParentHere.
 * 
 * @param $tpls_itemParentUnpub {string_chunkName|string} — The menu item template for unpublished documents which has a children displayed in menu. Default: $tpls_itemParent.
 * @param $tpls_itemParentUnpubActive {string_chunkName|string} — The menu item template for an unpublished document which has the current document as one of the children. Default: $tpls_itemParentActive.
 * 
 * @param $tpls_outer {string_chunkName|string} — Wrapper template. Available placeholders: [+children+]. Default: '<ul>[+children+]</ul>'.
 * 
 * @param $placeholders {stirng_json|string_queryFormated} — Additional data as query string has to be passed into “tpls_outer”. The parameter must be set as JSON (https://en.wikipedia.org/wiki/JSON) or Query string (https://en.wikipedia.org/wiki/Query_string). Default: —.
 * @example &placeholders=`{"pladeholder1": "value1", "pagetitle", "My awesome pagetitle!"}`.
 * @example &placeholders=`pladeholder1=value1&pagetitle=My awesome pagetitle!`.
 * 
 * @link http://code.divandesign.biz/modx/ddmenubuilder/1.13b
 * 
 * @copyright 2009–2018 DivanDesign {@link http://www.DivanDesign.biz }
 */

//Подключаем класс (ddTools подключится там)
require_once $modx->getConfig('base_path').'assets/snippets/ddMenuBuilder/ddMenuBuilder.class.php';

//Для обратной совместимости
extract(ddTools::verifyRenamedParams($params, [
	'tpls_item' => 'tplRow',
	'tpls_itemHere' => 'tplHere',
	'tpls_itemActive' => 'tplActive',
	'tpls_itemParent' => 'tplParentRow',
	'tpls_itemParentHere' => 'tplParentHere',
	'tpls_itemParentActive' => 'tplParentActive',
	'tpls_itemParentUnpub' => 'tplUnpubParentRow',
	'tpls_itemParentUnpubActive' => 'tplUnpubParentActive',
	'tpls_outer' => 'tplWrap'
]));

//Backward compatibility
if (
	isset($startId) &&
	is_numeric($startId)
){
	//По умолчанию на 1 уровень
	$providerParams = '{
		"parentIds": "'.$startId.'",
		"depth": '.(isset($depth) && is_numeric($depth) ? $depth : 1).'
	}';
}

$ddMenuBuilder_params = new stdClass();

//Задаём шаблоны
$ddMenuBuilder_params->templates = [];

if (isset($tpls_item)){$ddMenuBuilder_params->templates['item'] = $modx->getTpl($tpls_item);}
if (isset($tpls_itemHere)){$ddMenuBuilder_params->templates['itemHere'] = $modx->getTpl($tpls_itemHere);}
if (isset($tpls_itemActive)){$ddMenuBuilder_params->templates['itemActive'] = $modx->getTpl($tpls_itemActive);}

if (isset($tpls_itemParent)){$ddMenuBuilder_params->templates['itemParent'] = $modx->getTpl($tpls_itemParent);}
if (isset($tpls_itemParentHere)){$ddMenuBuilder_params->templates['itemParentHere'] = $modx->getTpl($tpls_itemParentHere);}
if (isset($tpls_itemParentActive)){$ddMenuBuilder_params->templates['itemParentActive'] = $modx->getTpl($tpls_itemParentActive);}

if (isset($tpls_itemParentUnpub)){$ddMenuBuilder_params->templates['itemParentUnpub'] = $modx->getTpl($tpls_itemParentUnpub);}
if (isset($tpls_itemParentUnpubActive)){$ddMenuBuilder_params->templates['itemParentUnpubActive'] = $modx->getTpl($tpls_itemParentUnpubActive);}

if (empty($ddMenuBuilder_params->templates)){unset($ddMenuBuilder_params->templates);}

$tpls_outer = (isset($tpls_outer)) ? $modx->getTpl($tpls_outer) : '<ul>[+children+]</ul>';

//Направление сортировки
if (isset($sortDir)){$ddMenuBuilder_params->sortDir = $sortDir;}
//По умолчанию будут только опубликованные документы
if (isset($showPublishedOnly)){$ddMenuBuilder_params->showPublishedOnly = $showPublishedOnly;}
//По умолчанию будут только документы, у которых стоит галочка «показывать в меню»
if (isset($showInMenuOnly)){$ddMenuBuilder_params->showInMenuOnly = $showInMenuOnly;}

$ddMenuBuilder = new ddMenuBuilder($ddMenuBuilder_params);

//Prepare provider params
$providerParams = ddTools::encodedStringToArray($providerParams);

//Генерируем меню
$result = $ddMenuBuilder->generate($ddMenuBuilder->prepareProviderParams([
	//Parent by default
	'provider' => isset($provider) ? $provider : 'parent',
	'providerParams' => $providerParams
]));

//Данные, которые необоходимо передать в шаблон
if (!empty($placeholders)){
	$placeholders = ddTools::encodedStringToArray($placeholders);
}else{
	$placeholders = [];
}

$placeholders['children'] = $result['outputString'];

return ddTools::parseText([
	'text' => $tpls_outer,
	'data' => $placeholders
]);
?>